### PR TITLE
8246348: Crash in libpango on Ubuntu 20.04 with some unicode chars

### PR DESCRIPTION
**Pull request by johanvos**
This addresses https://bugs.openjdk.java.net/browse/JDK-8246348

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8246348](https://bugs.openjdk.java.net/browse/JDK-8246348): Crash in libpango on Ubuntu 20.04 with some unicode chars

### Reviewers
 * Phil Race ([prr](@prrace) - **Reviewer**)
 * Kevin Rushforth ([kcr](@kevinrushforth) - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/249/head:pull/249`
`$ git checkout pull/249`

### CODE REVIEW

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java [hunk 1]
```diff
@@ -80,9 +84,11 @@ private boolean check(long checkValue, String message, long context, long desc,
         return true;
     }
 
-    private long str = 0L;
+    private Map<TextRun, Long> runUtf8 = new HashMap<>();
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:87]: In order for this to work, each TextRun would need to be immutable (at least during the life of this PangoGlyphLayout) and there would need to be a 1-1 mapping between the TextRun and the string pointer, independent of anything else. Is this the case? I'm seeing some crashes in OSPango.g_utf8_offset_to_pointer when testing. I don't know if they are related to this or not.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java [hunk 2]
```diff
@@ -80,9 +84,11 @@ private boolean check(long checkValue, String message, long context, long desc,
...
     public void layout(TextRun run, PGFont font, FontStrike strike, char[] text) {
-
+        for (char c: text) {
+            if (c == 0) c = '\f';
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:90]: This won't actually do anything (it just sets a local variable that immediately goes out of scope). I don't think that we want to modify the array itself, so you will probably want to make a copy of the array in the case there is a '0' character.

As for what to replace the '0' char with, maybe a space? @prrace can probably suggest something.

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:90]: Interestingly enough, I don't see a crash due to the 0 character, but I am seeing some new assertion warnings expecting `length >= 0` even in cases where there isn't a 0 character.

Comment (johanvos) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:90]: I saw that assertion as well, but that can come from cases where empty TextRun instances are used (start == end). In that case, the length is 0, which pango considers silly, but it won't crash on that (the content array is never examined in that case)

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java (hunk 1 above)
Comment (johanvos) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:87]: I would assume that once the layout method is called, the TextRun won't change anymore. If it would change, the current code might break as well, as the string pointer is currently cached.
I agree though that this is a brittle approach. What I'm basically doing here is store additional information (the number of unicode codepoints rather than the number of UTF 16 chars, which can then be mapped to the string pointer pointing to the relevant substring) on the TextRun, but since I didn't want to modify TextRun, that additional info is stored in this Map.
I'd prefer modifying TextRun though.

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:87]: I think I see what you are trying to do, but there are a couple of problems it is causing. I attached a new test program to the JBS bug report, UnicodeTextTest2.java, which avoids any null (0) characters. That test runs fine today, and with the first version of your fix (the one without the Map), but causes assertion failures with the current version of the patch:

```
(java:32761): Pango-CRITICAL **: 10:37:09.535: pango_itemize: assertion 'length >= 0' failed

(java:32761): Pango-CRITICAL **: 10:37:09.963: pango_itemize: assertion 'length >= 0' failed
```

I also sometimes get this:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f00456c2d78, pid=32761, tid=32778
#
# JRE version: Java(TM) SE Runtime Environment (14.0+36) (build 14+36-1461)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (14+36-1461, mixed mode, sharing, tiered, compressed oops, serial gc, linux-amd64)
# Problematic frame:
# C  [libglib-2.0.so.0+0x84d78]  g_utf8_offset_to_pointer+0x28
```

Phil is the expert here, so I'd like him to weigh in.

**Comment by prrace** [comment]
Instead of a space you could use a ZWNJ U+FEFF. Because that is also the endian-ness
mark, Unicode actually prefers you now to use U+2060 but you need to make sure these
are processed correctly by pango. Unlike a space they should have no rendering effect - not even an advance.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java [hunk 3]
```diff
@@ -126,16 +132,20 @@ public void layout(TextRun run, PGFont font, FontStrike strike, char[] text) {
...
             if (check(str, "Failed allocating UTF-8 buffer.", context, desc, attrList)) {
                 return;
             }
         }
 
         /* Itemize */
         long start = OSPango.g_utf8_offset_to_pointer(str, run.getStart());
-        long end = OSPango.g_utf8_offset_to_pointer(str, run.getEnd());
+        long utflen = OSPango.g_utf8_strlen(str,-1);
+        long end = OSPango.g_utf8_offset_to_pointer(str, utflen);
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:148]: Since you are now creating `n` native strings, 1 per substring based on the `TextRun`, rather than 1 for the entire String, isn't the `start` pointer wrong? Unless I am missing something, I would think that `start` should be set to `str`.

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:148]: I did a quick test, and setting `start = str` fixes the spurious assertions and intermittent crash.

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:148]: I should add that this is without any attempt to filter out `0` chars. Both `UnicodeTextTest` and  `UnicodeTextTest2` run correctly with no crashes and no assertions when I comment out the (ineffective) loop checking for 0 and set `start = str` leaving everything else as you have it in the current PR. Loading https://gluonhq.com/ in WebView works, too.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java (hunk 1 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:87]: Given my observation with the `start` pointer, I think the `HashMap` as you are using it is fine. Do you think a `LinkedHashMap` might be useful to preserve the iteration order? (since you only iterate it once in `dispose` it probably doesn't matter).

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java [hunk 4]
```diff
@@ -126,16 +132,20 @@ public void layout(TextRun run, PGFont font, FontStrike strike, char[] text) {
...
         }
 
-        if (str == 0L) {
-            str = OSPango.g_utf16_to_utf8(text);
+        Long str = runUtf8.get(run);
+        if (str == null) {
+            char[] rtext = Arrays.copyOfRange(text, run.getStart(), run.getEnd());
+            str = OSPango.g_utf16_to_utf8(rtext);
+            runUtf8.put(run, str);
             if (check(str, "Failed allocating UTF-8 buffer.", context, desc, attrList)) {
```
Comment (prrace) [inline modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:140]: Did you really mean to store the run in the map before checking if str is null ?

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java (hunk 3 above)
Comment (prrace) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:148]: I agree. the text array is now the subset for the run, so effectively run.getStart() should be zero,
so there should be no need for that call. This probably looks OK when there is one run not otherwise.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java (hunk 4 above)
Comment (johanvos) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:140]: I'm not sure what should happen if str is null. Do we want to be able to retry later? In that case, we shouldn't store it in the map. But if it is fatal, and seems to indicate a problem with rtext (e.g. too big?), so in that case it might not be worth retrying.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java (hunk 2 above)
Comment (johanvos) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:90]: the crash with 0 is gone with the second patch indeed (1 str per Run), which might be a consequence of doing strlen calculations before we send content down to the pango lib

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java (hunk 3 above)
Comment (johanvos) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:144]: You're both correct. I simplified it.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java (hunk 4 above)
Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:137]: If `str` is null, you would end up retrying it anyway, since `Map::get` will return `null` the next time (if there is a next time). Also, I think the dispose method might then crash trying to free a null pointer. So I think Phil is right and that you should not add the `str` to the `Map` if it is null.

**Comment by kevinrushforth** [comment]
Once you address the question of storing a null value in the Map, the only remaining question I have is whether to use a LinkedHashMap instead of an ordinary HashMap. In general, I like the predictability of a LinkedHashMap for maps that are iterated, but in this case, I don't feel strongly about it one way or the other.

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java (hunk 4 above)
Comment (johanvos) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:137]: I don't think it will be `null` but it will be `0` in which case it is stored in the `Map`
The only reason the str == null is then that we never tried to convert the Java chars to utf8, unless I'm missing a case?

Comment (kevinrushforth) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:137]: Oh, right. It will be a Long 0, not null. If you store it you will still have the problem I mentioned with dispose unless you add back in the `str != 0` check. And you would need a check for `str != 0` in the layout method so that the second time it doesn't treat it as a valid pointer. It might be safer to not store it if 0, which matches the current behavior?

Comment (johanvos) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:136]: That makes sense indeed. For the dispose call, a valid pointer is required.

**Comment by kevinrushforth** [comment]
The fix looks good now.

As for the test, even if StubFont were updated to provide a real font, the StubToolkit doesn't load Prism (so none of the text rendering code is exercised). Do you think you could instead add a simple test or two in the system tests project instead (maybe one testing UTF16 chars and one with a 0 char)?

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java (hunk 4 above)
Comment (prrace) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:136]: I wasn't completely sure what the idea was, which is why I phrased it as a question :-)

File: modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java (hunk 3 above)
Comment (prrace) [reply modules/javafx.graphics/src/main/java/com/sun/javafx/font/freetype/PangoGlyphLayout.java:144]: Looks good now.

**Comment by johanvos** [comment]
> As for the test, even if StubFont were updated to provide a real font, the StubToolkit doesn't load Prism (so none of the text rendering code is exercised). Do you think you could instead add a simple test or two in the system tests project instead (maybe one testing UTF16 chars and one with a 0 char)?

I added a systemtest

File: tests/system/src/test/java/test/com/sun/javafx/font/freetype/PangoTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        @Override
+        public void start(Stage primaryStage) throws Exception {
+            this.stage = primaryStage;
+            pane = new VBox(10);
+            Scene scene = new Scene(pane, 400, 200);
+            stage.setScene(scene);
+            stage.show();
+            launchLatch.countDown();
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/com/sun/javafx/font/freetype/PangoTest.java:77]: Our newer system tests use a `WINDOW_SHOWN` event to trigger the countdown of the latch, like this (which would be added before the stage is shown).
```
            stage.addEventHandler(WindowEvent.WINDOW_SHOWN, e -> Platform.runLater(launchLatch::countDown));
```

This ensures that the Stage has been shown before the first test is run.

File: tests/system/src/test/java/test/com/sun/javafx/font/freetype/PangoTest.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        Platform.runLater(() -> {
+            text.layoutYProperty().addListener(inv -> {
+                rDone.countDown();
+            });
+            pane.getChildren().add(text);
+        });
+
+        try {
+            if (!rDone.await(TIMEOUT, TimeUnit.MILLISECONDS)) {
+                throw new AssertionFailedError("Timeout waiting for runLater");
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/com/sun/javafx/font/freetype/PangoTest.java:117]: This could also be replaced with `assertTrue("Timeout ...", rDone.await(TIMEOUT, TimeUnit.MILLISECONDS)` if you add `throws Exception` to this method and the calling test methods.

File: modules/javafx.graphics/src/test/java/test/com/sun/javafx/text/TextLayoutTest.java [hunk 7]
```diff
@@ -95,6 +98,21 @@ private void verifyComplex(PrismTextLayout layout, boolean... complex) {
         }
     }
 
+    @Ignore() // ignored since StubFontLoader used in tests return fonts with null resources
+    @Test public void utf16chars() {
```
Comment (kevinrushforth) [inline modules/javafx.graphics/src/test/java/test/com/sun/javafx/text/TextLayoutTest.java:102]: Since this is `@Ignore`d with no real way to make it a useful test, maybe it would be best to revert the changes to this file?

File: tests/system/src/test/java/test/com/sun/javafx/font/freetype/PangoTest.java [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    @BeforeClass
+    public static void setupOnce() {
+        // Start the Application
+        new Thread(() -> Application.launch(MyApp.class, (String[]) null)).start();
+
+        try {
+            if (!launchLatch.await(TIMEOUT, TimeUnit.MILLISECONDS)) {
+                throw new AssertionFailedError("Timeout waiting for Application to launch");
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/com/sun/javafx/font/freetype/PangoTest.java:88]: Somehow this comment was dropped from my previous comments...

If you add `throws Exception` to this method, then the entire try/catch block can be replaced with:
```
        assertTrue("Timeout waiting for Application to launch", launchLatch.await(TIMEOUT, TimeUnit.MILLISECONDS));
```
